// hermes-tree-select/lib/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_ROOT = exports.toggle = exports.DEFAULT_NO_FOUND_CONTENT = exports.DEFAULT_ASYNC_NODE_TEXT = exports.DEFAULT_SYNC_NODE_TEXT = undefined;

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _icon = require('antd/lib/icon');

var _icon2 = _interopRequireDefault(_icon);

require('antd/lib/icon/style/css');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_SYNC_NODE_TEXT = exports.DEFAULT_SYNC_NODE_TEXT = function DEFAULT_SYNC_NODE_TEXT(node, isRight) {
  var rtn = false;
  if (node) {
    if (isRight) {
      if (node.id === '#') {
        rtn = _react2.default.createElement(
          'span',
          null,
          '\u5DF2\u9009 ',
          '' + node.checked().length
        );
      } else if (node.isLeaf()) {
        rtn = _react2.default.createElement(
          'span',
          null,
          node.model.name
        );
      } else {
        rtn = _react2.default.createElement(
          'span',
          null,
          node.model.name,
          '(',
          node.checked().length,
          ')'
        );
      }
    } else if (node.id === '#') {
      rtn = _react2.default.createElement(
        'span',
        null,
        '\u5171 ',
        '' + node.leafs().length
      );
    } else if (node.isLeaf()) {
      rtn = _react2.default.createElement(
        'span',
        null,
        node.model.name
      );
    } else {
      rtn = _react2.default.createElement(
        'span',
        null,
        node.model.name,
        '(',
        node.leafs().length,
        ')'
      );
    }
  } else {
    rtn = _react2.default.createElement(
      'span',
      null,
      '\u641C\u7D22\u7ED3\u679C\u4E3A\u7A7A'
    );
  }
  return rtn;
};

var DEFAULT_ASYNC_NODE_TEXT = exports.DEFAULT_ASYNC_NODE_TEXT = function DEFAULT_ASYNC_NODE_TEXT(node, isRight) {
  var rtn = false;
  if (node) {
    if (isRight) {
      if (node.id === '#') {
        rtn = _react2.default.createElement(
          'span',
          null,
          '\u5DF2\u9009 ',
          '' + node.checked().length
        );
      } else if (node.isLeaf()) {
        rtn = _react2.default.createElement(
          'span',
          null,
          node.model.name
        );
      } else {
        rtn = _react2.default.createElement(
          'span',
          null,
          node.model.name,
          '(',
          node.checked().length,
          ')'
        );
      }
    } else if (node.id === '#') {
      rtn = '全选';
    } else if (node.isLeaf()) {
      rtn = _react2.default.createElement(
        'span',
        null,
        node.model.name
      );
    } else {
      rtn = _react2.default.createElement(
        'span',
        null,
        node.model.name,
        '(',
        node.degree(),
        ')'
      );
    }
  } else if (!isRight) {
    rtn = _react2.default.createElement(
      'span',
      null,
      '\u641C\u7D22\u7ED3\u679C\u4E3A\u7A7A'
    );
  } else {
    rtn = _react2.default.createElement(_icon2.default, { type: 'loading' });
  }
  return rtn;
};

var DEFAULT_NO_FOUND_CONTENT = exports.DEFAULT_NO_FOUND_CONTENT = function DEFAULT_NO_FOUND_CONTENT() {
  return _react2.default.createElement(
    'div',
    { className: 'no-leaf' },
    _react2.default.createElement(_icon2.default, { type: 'meh' }),
    ' Not Found'
  );
};

var toggle = exports.toggle = function toggle(array, value) {
  var rtn = [].concat((0, _toConsumableArray3.default)(array));
  var index = rtn.indexOf(value);
  if (index === -1) {
    rtn.push(value);
  } else {
    rtn.splice(index, 1);
  }
  return rtn;
};

var DEFAULT_ROOT = exports.DEFAULT_ROOT = {
  id: '#',
  name: 'ROOT',
  count: 1e5
};
