// hermes-react/lib/info-modal/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _SampleModal = require('./SampleModal');

var _SampleModal2 = _interopRequireDefault(_SampleModal);

require('./style/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

function noop() {}

var InfoModal = (_temp = _class = function (_Component) {
  _inherits(InfoModal, _Component);

  function InfoModal(props) {
    _classCallCheck(this, InfoModal);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.showModal = function (e) {
      e.preventDefault();
      _this.setState({
        visible: true
      });
    };

    _this.hideModal = function () {
      _this.setState({
        visible: false
      });
      _this.props.onOk();
    };

    _this.state = {
      visible: false
    };
    return _this;
  }

  InfoModal.prototype.render = function render() {
    var _props = this.props,
        trigger = _props.trigger,
        style = _props.style,
        className = _props.className,
        others = _objectWithoutProperties(_props, ['trigger', 'style', 'className']);

    var newStyle = _extends({
      display: 'inline-block'
    }, style);

    var props = _extends({}, others, {
      visible: this.state.visible,
      onOk: this.hideModal,
      onCancel: this.hideModal
    });

    return _react2["default"].createElement(
      'div',
      { style: newStyle, className: className },
      _react2["default"].createElement(
        'a',
        { onClick: this.showModal },
        trigger
      ),
      _react2["default"].createElement(
        _SampleModal2["default"],
        props,
        this.props.children
      )
    );
  };

  return InfoModal;
}(_react.Component), _class.propTypes = {
  style: _propTypes2["default"].object,
  trigger: _propTypes2["default"].any,
  children: _propTypes2["default"].any,
  className: _propTypes2["default"].string,
  onOk: _propTypes2["default"].func,
  okText: _propTypes2["default"].string
}, _class.defaultProps = {
  trigger: '查看',
  title: '查看',
  onOk: noop
}, _temp);
exports["default"] = InfoModal;
module.exports = exports['default'];
