// hermes-react/lib/lazy-img/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

var _container = require('./container');

var _container2 = _interopRequireDefault(_container);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var LazyImg = (_temp = _class = function (_Base) {
  _inherits(LazyImg, _Base);

  function LazyImg() {
    _classCallCheck(this, LazyImg);

    return _possibleConstructorReturn(this, _Base.apply(this, arguments));
  }

  LazyImg.prototype.componentDidMount = function componentDidMount() {
    var containerComponent = this.props.container;
    if (containerComponent) {
      this.isWindowMode = false;
      containerComponent.addImage((0, _reactDom.findDOMNode)(this));
      containerComponent.reinit();
    } else {
      this.images = [(0, _reactDom.findDOMNode)(this)];
      var container = this.findContainer(this.images[0]);
      if (!container) {
        this.container = window;
        this.isWindowMode = true;
        this.bind();
      } else {
        this.isWindowMode = false;
      }
    }
  };

  LazyImg.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
    if (this.props.src !== prevProps.src) {
      var image = (0, _reactDom.findDOMNode)(this);
      image.setAttribute('src', this.props.placeholder);
      if (this.isWindowMode) {
        this.addImage(image);
        this.rebind();
      } else {
        var containerComponent = this.props.container;
        if (containerComponent) {
          containerComponent.addImage(image);
          containerComponent.reinit();
        }
      }
    }
  };

  LazyImg.prototype.componentWillUnmount = function componentWillUnmount() {
    if (this.isWindowMode) {
      this.unbind();
    }
  };

  LazyImg.prototype.addImage = function addImage(image) {
    if (this.images.indexOf(image) === -1) {
      this.images.push(image);
    }
  };

  LazyImg.prototype.render = function render() {
    var _props = this.props,
        src = _props.src,
        placeholder = _props.placeholder,
        offset = _props.offset;

    var props = _extends({}, this.props);
    ['key', 'container', 'src', 'placeholder', 'offset'].forEach(function (v) {
      return delete props[v];
    });
    return _react2["default"].createElement('img', _extends({}, props, {
      src: placeholder,
      'data-lazy-img': true,
      'data-src': src,
      'data-offset': offset
    }));
  };

  return LazyImg;
}(_base2["default"]), _class.propTypes = {
  src: _propTypes2["default"].string,
  placeholder: _propTypes2["default"].string,
  offset: _propTypes2["default"].number,
  container: _propTypes2["default"].object
}, _class.defaultProps = {
  src: '',
  placeholder: 'https://gw.alipayobjects.com/zos/rmsportal/WlzpIBsCaFAawKKaQWRV.png',
  offset: 0,
  container: null
}, _class.Container = _container2["default"], _temp);
exports["default"] = LazyImg;
module.exports = exports['default'];
