// hermes-react/lib/img-slider-modal/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _modal = require('antd/lib/modal');

var _modal2 = _interopRequireDefault(_modal);

var _carousel = require('antd/lib/carousel');

var _carousel2 = _interopRequireDefault(_carousel);

var _icon = require('antd/lib/icon');

var _icon2 = _interopRequireDefault(_icon);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _class, _temp;

require('antd/lib/modal/style/css');

require('antd/lib/carousel/style/css');

require('antd/lib/icon/style/css');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

require('./style/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var PrevArrow = function (_Component) {
  _inherits(PrevArrow, _Component);

  function PrevArrow() {
    _classCallCheck(this, PrevArrow);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  PrevArrow.prototype.render = function render() {
    var onClick = this.props.onClick;

    return _react2["default"].createElement(
      'div',
      { className: 'arrow-prev' + (!this.props.onClick ? ' arrow-disabled' : ''), onClick: onClick },
      _react2["default"].createElement(_icon2["default"], { type: 'left' })
    );
  };

  return PrevArrow;
}(_react.Component);

var NextArrow = function (_Component2) {
  _inherits(NextArrow, _Component2);

  function NextArrow() {
    _classCallCheck(this, NextArrow);

    return _possibleConstructorReturn(this, _Component2.apply(this, arguments));
  }

  NextArrow.prototype.render = function render() {
    var onClick = this.props.onClick;

    return _react2["default"].createElement(
      'div',
      { className: 'arrow-next' + (!this.props.onClick ? ' arrow-disabled' : ''), onClick: onClick },
      _react2["default"].createElement(_icon2["default"], { type: 'right' })
    );
  };

  return NextArrow;
}(_react.Component);

var ImgSliderModal = (_temp = _class = function (_Component3) {
  _inherits(ImgSliderModal, _Component3);

  function ImgSliderModal(props) {
    _classCallCheck(this, ImgSliderModal);

    var _this3 = _possibleConstructorReturn(this, _Component3.call(this, props));

    _this3.showModal = function () {
      _this3.setState({
        showModal: true
      });
    };

    _this3.hideModal = function () {
      _this3.setState({
        showModal: false
      });
    };

    _this3.state = {
      showModal: false
    };
    return _this3;
  }

  ImgSliderModal.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        children = _props.children,
        list = _props.list,
        width = _props.width,
        height = _props.height,
        dots = _props.dots;
    var showModal = this.state.showModal;

    var simpleMode = list && list.length && typeof list[0] === 'string';
    return children ? _react2["default"].createElement(
      'div',
      { className: 'hermes-imageslider' + (className ? ' ' + className : ''), style: style },
      list.length && showModal ? _react2["default"].createElement(
        _modal2["default"],
        { className: 'hermes-imageslider-modal', width: width + 32 + 'px', visible: true, maskClosable: false, footer: null, onCancel: this.hideModal },
        _react2["default"].createElement(
          _carousel2["default"],
          _extends({}, this.props, {
            className: 'hermes-imageslider-carousel' + (!dots ? ' no-padding' : ''),
            style: null,
            prevArrow: _react2["default"].createElement(PrevArrow, null),
            nextArrow: _react2["default"].createElement(NextArrow, null)
          }),
          list.map(function (v, i) {
            return _react2["default"].createElement(
              'div',
              { className: 'hermes-imagewrap', style: { height: height + 'px', lineHeight: height + 'px' }, key: i },
              _react2["default"].createElement('img', { alt: simpleMode ? '' : v.alt, style: { maxHeight: height + 'px', maxWidth: width + 'px' }, src: simpleMode ? v : v.src }),
              !simpleMode && 'desc' in v && v.desc !== '' ? _react2["default"].createElement(
                'div',
                null,
                v.desc
              ) : null
            );
          })
        )
      ) : null,
      _react2["default"].createElement(
        'div',
        { className: 'hermes-imageslider-trigger', onClick: this.showModal },
        children
      )
    ) : null;
  };

  return ImgSliderModal;
}(_react.Component), _class.propTypes = {
  className: _propTypes2["default"].string,
  style: _propTypes2["default"].object,
  children: _propTypes2["default"].any,
  imgList: _propTypes2["default"].array,
  width: _propTypes2["default"].number,
  height: _propTypes2["default"].number,
  arrows: _propTypes2["default"].bool,
  dots: _propTypes2["default"].bool,
  infinite: _propTypes2["default"].bool
}, _class.defaultProps = {
  children: null,
  imgList: [],
  width: 688,
  height: 500,
  arrows: true,
  dots: true,
  infinite: true
}, _temp);
exports["default"] = ImgSliderModal;
module.exports = exports['default'];
