// hermes-react/lib/tag-select/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _Common = require('./Common');

var _Common2 = _interopRequireDefault(_Common);

require('./style/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var TagSelect = (0, _createReactClass2["default"])({
  displayName: 'TagSelect',

  mixins: [_Common2["default"]],
  getInitialState: function getInitialState() {
    var _props = this.props,
        limit = _props.limit,
        limitedDisable = _props.limitedDisable;

    var _getState = this.getState(),
        tagList = _getState.tagList,
        selected = _getState.selected;

    return {
      tagList: tagList,
      selected: selected,
      disableAll: limitedDisable && selected.length >= limit
    };
  }
});

TagSelect.RadioSelect = (0, _createReactClass2["default"])({
  displayName: 'RadioSelect',

  mixins: [_Common2["default"]],
  getInitialState: function getInitialState() {
    var _getState2 = this.getState(),
        tagList = _getState2.tagList,
        selected = _getState2.selected;

    return {
      tagList: tagList,
      selected: selected,
      radioMode: true
    };
  }
});

exports["default"] = TagSelect;
module.exports = exports['default'];
