// hermes-react/lib/promo-bar/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _class, _temp, _initialiseProps;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

require('./style/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var instanceCount = 0;

var PromoBar = (_temp = _class = function (_Component) {
  _inherits(PromoBar, _Component);

  function PromoBar(props) {
    _classCallCheck(this, PromoBar);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _initialiseProps.call(_this);

    var dataSource = props.dataSource;

    var dataDoms = dataSource.map(function (v, i, a) {
      var custom = v.custom,
          link = v.link,
          content = v.content,
          name = v.name;

      var hasLink = !!link;
      var dom = custom ? _react2["default"].createElement(
        'span',
        null,
        custom
      ) : _react2["default"].createElement(
        'span',
        null,
        i + 1,
        '.\u3010',
        name,
        '\u3011\uFF1A',
        content
      );
      return hasLink ? _react2["default"].createElement(
        'a',
        {
          style: a.length === 1 && i === 0 ? { marginLeft: 0 } : null,
          key: i,
          className: 'hermes-promobar-item hermes-promobar-has-link',
          href: link,
          target: '_blank',
          rel: 'noopener noreferrer',
          title: content },
        dom
      ) : _react2["default"].createElement(
        'span',
        {
          style: a.length === 1 && i === 0 ? { marginLeft: 0 } : null,
          key: i,
          className: 'hermes-promobar-item' },
        dom
      );
    });
    _this.state = {
      isShow: true,
      dataDoms: dataDoms,
      inlineStyle: null
    };
    return _this;
  }

  PromoBar.prototype.componentDidMount = function componentDidMount() {
    var dataDoms = this.state.dataDoms;

    if (dataDoms.length) {
      this.instance = instanceCount++;
      this.init();
    }
  };

  PromoBar.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        className = _props.className,
        style = _props.style,
        closable = _props.closable,
        vertical = _props.vertical;
    var _state = this.state,
        isShow = _state.isShow,
        dataDoms = _state.dataDoms,
        inlineStyle = _state.inlineStyle;
    var startScroll = this.startScroll,
        startScrollV = this.startScrollV,
        stopScroll = this.stopScroll,
        stopScrollV = this.stopScrollV,
        onHide = this.onHide;

    var content = vertical ? _react2["default"].createElement(
      'div',
      { className: 'hermes-promobar-list-v', style: inlineStyle },
      dataDoms
    ) : _react2["default"].createElement(
      'div',
      { className: 'hermes-promobar-list hermes-promobar-animate-' + this.instance, style: inlineStyle },
      dataDoms
    );
    return isShow ? _react2["default"].createElement(
      'div',
      { className: 'hermes-promobar' + (className ? ' ' + className : ''), style: style },
      closable ? _react2["default"].createElement(
        'span',
        { className: 'hermes-promobar-close' },
        _react2["default"].createElement(
          'a',
          { onClick: onHide },
          '\xD7'
        )
      ) : null,
      _react2["default"].createElement(
        'div',
        {
          className: 'hermes-promobar-container',
          ref: function ref(_ref) {
            return _this2.hermesPromoContainer = _ref;
          },
          onMouseEnter: vertical ? stopScrollV : stopScroll,
          onMouseLeave: vertical ? startScrollV : startScroll },
        content
      ),
      _react2["default"].createElement('style', { ref: function ref(_ref2) {
          return _this2.stylesheet = _ref2;
        } })
    ) : null;
  };

  return PromoBar;
}(_react.Component), _class.propTypes = {
  className: _propTypes2["default"].string,
  style: _propTypes2["default"].object,
  dataSource: _propTypes2["default"].array,
  closable: _propTypes2["default"].bool,
  onClose: _propTypes2["default"].func,
  speed: _propTypes2["default"].number,
  vertical: _propTypes2["default"].bool,
  duration: _propTypes2["default"].number
}, _class.defaultProps = {
  dataSource: [],
  closable: true,
  onClose: function onClose() {},
  speed: 50,
  vertical: false,
  duration: 5
}, _initialiseProps = function _initialiseProps() {
  var _this3 = this;

  this.onHide = function () {
    var onClose = _this3.props.onClose;

    _this3.setState({
      isShow: false
    }, onClose);
  };

  this.init = function () {
    var vertical = _this3.props.vertical;

    if (vertical) {
      _this3.fillPromBarV();
      if (_this3.isScroll) _this3.startScrollV();
    } else {
      _this3.fillPromBar();
      if (_this3.isScroll) _this3.startScroll();
    }
  };

  this.fillPromBar = function () {
    var _props2 = _this3.props,
        speed = _props2.speed,
        dataSource = _props2.dataSource;
    var dataDoms = _this3.state.dataDoms;

    var promoContainer = _this3.hermesPromoContainer;
    var containerW = promoContainer.offsetWidth;
    var listW = promoContainer.children[0].offsetWidth;
    if (dataSource.length === 1 && listW < containerW) {
      _this3.isScroll = false;
      return;
    }
    _this3.isScroll = true;
    var repeatTimes = Math.ceil(containerW / listW);
    var dataDomsRepeat = Array.prototype.slice.call({ length: repeatTimes + 1 }).join(' ').split(' ').map(function () {
      return dataDoms;
    });
    var duration = (listW / speed).toFixed(1);
    _this3.stylesheet.innerHTML = '\n      .hermes-promobar-animate-' + _this3.instance + ' {\n        animation: hermes-promobar-scroll-' + _this3.instance + ' ' + duration + 's infinite linear;\n        animation-play-state: paused;\n      }\n      @keyframes hermes-promobar-scroll-' + _this3.instance + ' {\n        from { transform: translate(0, 0); }\n        to { transform: translate(-' + listW + 'px, 0); }\n      }\n    ';
    _this3.setState({
      dataDoms: dataDomsRepeat
    });
  };

  this.fillPromBarV = function () {
    var dataSource = _this3.props.dataSource;
    var dataDoms = _this3.state.dataDoms;

    if (dataSource.length === 1) {
      _this3.isScroll = false;
      return;
    }
    dataDoms.push(dataDoms[0]);
    var dataDomsRepeat = dataDoms.map(function (v, i) {
      return _react2["default"].createElement(
        'div',
        { className: 'hermes-promobar-item-wrap', key: i },
        v
      );
    });
    _this3.isScroll = true;
    _this3.setState({
      dataDoms: dataDomsRepeat
    });
  };

  this.stopScroll = function () {
    if (!_this3.isScroll) return;
    _this3.setState({
      inlineStyle: { animationPlayState: 'paused' }
    });
  };

  this.startScroll = function () {
    if (!_this3.isScroll) return;
    _this3.setState({
      inlineStyle: { animationPlayState: 'running' }
    });
  };

  this.stopScrollV = function () {
    if (!_this3.isScroll) return;
    clearInterval(_this3.i);
    _this3.i = null;
  };

  this.startScrollV = function () {
    var _props3 = _this3.props,
        duration = _props3.duration,
        dataSource = _props3.dataSource;

    var promoContainer = _this3.hermesPromoContainer;
    var containerH = promoContainer.offsetHeight;
    _this3.offset = _this3.offset || 0;
    if (!_this3.isScroll) return;
    var scrollFn = function scrollFn() {
      _this3.offset = _this3.offset - containerH;
      _this3.setState({
        inlineStyle: { transform: 'translateY(' + _this3.offset + 'px)' }
      });
    };
    _this3.i = setInterval(function () {
      if (_this3.offset <= -containerH * dataSource.length) {
        _this3.offset = 0;
        _this3.setState({
          inlineStyle: { transition: 'none 0s ease 0s' }
        });
        setTimeout(scrollFn, 20);
      } else {
        scrollFn();
      }
    }, (duration > 2 ? duration : 2) * 1000);
  };
}, _temp);
exports["default"] = PromoBar;
module.exports = exports['default'];
